dll/win32/shell32/CDefViewUtil.cpp

```diff
@@ -0,0 +1,132 @@
+        LPFNVIEWCALLBACK        m_Callback;
+        FOLDERVIEWMODE          m_FVM;
+        LONG                    m_Events;
+    public:
+        CShellFolderViewCBWrapper() : m_hWndMain(NULL) {}
+
+        HRESULT WINAPI Initialize(LPCSFV psvcbi)
+        {
+            m_psf = psvcbi->pshf;
+            m_psvOuter = psvcbi->psvOuter;
+            m_Callback = psvcbi->pfnCallback;
+            m_FVM = psvcbi->fvm;
+            m_Events = psvcbi->lEvents;
+            return S_OK;
+        }
+
+        // IShellFolderViewCB
+        STDMETHODIMP MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam) override
+        {
+            switch (uMsg)
+            {
+                case SFVM_HWNDMAIN:
+                    m_hWndMain = (HWND)lParam;
+                    break;
+
+                case SFVM_DEFVIEWMODE:
+                    if (m_FVM)
+                        *(FOLDERVIEWMODE*)lParam = m_FVM;
+                    break;
+            }
+
+            HRESULT hr = m_Callback(m_psvOuter, m_psf, m_hWndMain, uMsg, wParam, lParam);
+            if (SUCCEEDED(hr))
+                return hr;
+
+            switch (uMsg)
+            {
+                case SFVM_GETNOTIFY:
+                    *(LPITEMIDLIST*)wParam = NULL; // TODO: Where is this pidl supposed to come from?
+                    *(LONG*)lParam = m_Events;
+                    return S_OK;
+            }
+            return hr;
+        }
+
+        // IObjectWithSite
+        STDMETHODIMP SetSite(IUnknown *pUnkSite) override
+        {
+            //learn.microsoft.com/en-us/windows/win32/shell/sfvm-setisfv
+            HRESULT hr = CObjectWithSiteBase::SetSite(pUnkSite);
+            MessageSFVCB(SFVM_SETISFV, 0, (LPARAM)pUnkSite);
+            return hr;
+        }
+
+        DECLARE_NO_REGISTRY()
+        DECLARE_NOT_AGGREGATABLE(CShellFolderViewCBWrapper)
+        BEGIN_COM_MAP(CShellFolderViewCBWrapper)
+            COM_INTERFACE_ENTRY_IID(IID_IShellFolderViewCB, IShellFolderViewCB)
+            COM_INTERFACE_ENTRY_IID(IID_IObjectWithSite, IObjectWithSite)
+        END_COM_MAP()
+};
+
+/*************************************************************************
+ * SHCreateShellFolderViewEx [SHELL32.174] (Win95+)
+ */
+EXTERN_C HRESULT WINAPI
+SHCreateShellFolderViewEx(_In_ LPCSFV pcsfv, _Out_ IShellView **ppsv)
+{
+    if (!ppsv)
+        return E_INVALIDARG;
+    *ppsv = NULL;
+
+    TRACE("sf=%p pidl=%p cb=%p mode=0x%08x outer=%p\n",
+          pcsfv->pshf, pcsfv->pidl, pcsfv->pfnCallback,
+          pcsfv->fvm, pcsfv->psvOuter);
+
+    CComPtr<IShellFolderViewCB> psfvcb;
+    SFV_CREATE create = { sizeof(create), pcsfv->pshf, pcsfv->psvOuter };
+
+    if (pcsfv->pfnCallback)
+    {
+        HRESULT hr = ShellObjectCreatorInit<CShellFolderViewCBWrapper>(pcsfv,
+                        IID_PPV_ARG(IShellFolderViewCB, &psfvcb));
+        if (FAILED(hr))
+            return hr;
+        create.psfvcb = psfvcb;
+    }
+    return SHCreateShellFolderView(&create, ppsv);
```

Comment:
I would expect the non-Ex function to call the (new??) Ex function; not the other way around.

Reply:
Yes one would expect that but this case in special. The Ex version is the original and was undocumented pre-DOJ (ordinal only, I don't know what its internal name was in 95). The non-Ex version is probably newer (higher ordinal number), at least it was not exported before IE4.

The Ex version suffers from a fatal flaw, there is no easy way to get your context data passed to your callback function. Perhaps it can be done with COM aggregation and psfOuter but that would be annoying. The non-Ex versjon forces you to implement your own COM object and so data access is easy.

MSDN recommends that everyone use the non-Ex version and 99% of everything in ROS already does.

Reply:
OK thanks for your explanation. Maybe this could be part of a function code comment for `SHCreateShellFolderViewEx`.

Reply:
There is already a short comment in here calling the function callback "legacy".